src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.cs

```diff
@@ -25,35 +40,54 @@ internal sealed partial class VisualStudioRuleSetManager : IWorkspaceService, ID
             _listener = listener;
         }
 
-        public IRuleSetFile GetOrCreateRuleSet(string ruleSetFileFullPath)
+        public IReferenceCountedDisposable<IRuleSetFile> GetOrCreateRuleSet(string ruleSetFileFullPath)
         {
-            if (!_ruleSetFileMap.TryGetValue(ruleSetFileFullPath, out var ruleSetFile))
+            ReferenceCountedDisposable<RuleSetFile> disposable = null;
+
+            lock (_gate)
             {
-                ruleSetFile = new RuleSetFile(ruleSetFileFullPath, _fileChangeService, this);
-                _ruleSetFileMap.Add(ruleSetFileFullPath, ruleSetFile);
+                // If we already have one in the map to hand out, great
+                if (_ruleSetFileMap.TryGetValue(ruleSetFileFullPath, out var weakReference))
+                {
+                    disposable = weakReference.TryAddReference();
+                }
+
+                if (disposable == null)
+                {
+                    // We didn't easily get a disposable, so one of two things is the case:
+                    //
+                    // 1. We have no entry in _ruleSetFileMap at all for this.
+                    // 2. We had an entry, but it was disposed and is no longer valid.
+
+                    // In either case, we'll create a new rule set file and add it to the map.
+                    disposable = new ReferenceCountedDisposable<RuleSetFile>(new RuleSetFile(ruleSetFileFullPath, this));
+                    _ruleSetFileMap.Add(ruleSetFileFullPath, new ReferenceCountedDisposable<RuleSetFile>.WeakReference(disposable));
```

Comment:
Can this throw `Argument Exception “Item with Same Key has already been added”` for case 2. above?

Reply:
Fixed.